system/include/webgpu/webgpu_cpp.h

```diff
@@ -1214,6 +1216,24 @@ namespace wgpu {
     Bool GetInstanceFeatures(InstanceFeatures * features);
     Proc GetProcAddress(Device device, char const * procName);
 
+    struct AdapterInfo {
+        AdapterInfo() = default;
+        ~AdapterInfo();
+        AdapterInfo(const AdapterInfo&) = delete;
+        AdapterInfo& operator=(const AdapterInfo&) = delete;
+        AdapterInfo(AdapterInfo&&);
+        AdapterInfo& operator=(AdapterInfo&&);
+        ChainedStructOut  * nextInChain = nullptr;
+        char const * const vendor = nullptr;
+        char const * const architecture = nullptr;
+        char const * const device = nullptr;
+        char const * const description = nullptr;
```

Comment:
Would it make sense to have fixed length strings here?  That way you could use some kind of static buffer on the C/C++ side and JS could fill in the values without needing to call malloc?

Reply:
Alternatively if you want to keep this mechanism, then a common way to do this would be to have JS-reused the string buffer to consecutive calls (e.g. use `realloc` of a global pointer), so that no `free` is ever needed (since successive calls don't allocate).

Reply:
Thanks for the review @sbc100!

1. I'm not sure about the fixed length strings here. This info can't change for the life the adapter but it's not clear to me how to implement this suggestion.

2. I didn't find any `_realloc()` call in Emscripten ;(
Where would you store the global pointer? Something like `LibraryWebGPU.$WebGPU.vendorPtr` where `malloc` would be called in the first place and then realloc?

Reply:
Re fixed length I was thinking to use arrays rather than pointers. e.g:

`char vendor[MAX_INFO_STR];`

That way no separate allocations are needed.

Regarding `_realloc`, you would just need to add it to the `__deps` of the JS function in order to use it.    Yes you could use something like `vendorPtr`.   You could also try to put all the strings into one single buffer with a single allocation.   Sounds like fixed arrays would be simpler though.

Reply:
@kainino0x Any preference?

Reply:
In standardization for webgpu.h, we chose to use allocated strings instead of fixed-length char arrays (IIRC - it seems I forgot to take notes on the topic in https://github.com/webgpu-native/webgpu-headers/issues/143). We could pick a max length and use char arrays for vendor/architecture/device, but it's hard to pick a length for `description` since it can be pretty long - for example in GL, layered implementations like ANGLE typically wrap their own description around the underlying implementation's description, so the string gets longer with each layer. It's a little unfortunate for emscripten because JS will almost never expose a description at all in WebGPU, but in the general case it makes sense. We also have other entry points that will use the same pattern to return arrays of enums, etc., that we probably don't want to have to stabilize fixed max sizes for, so it'll be a pattern.

Reply:
We also didn't want to have pointers invalidated subtly by other calls, as would happen if we made them live until next overwrite. For that, plus reasons relating to implementing webgpu.h in rust, we also didn't want to require the strings to be owned by the parent object either (even if they would be immutable).

Reply:
Thank you @kainino0x for the background.

I've updated this PR with `realloc`. Let me know if that looks good to you

Reply:
If it might be of help, in juj/wasm_webgpu I implemented `wgpu_adapter_get_info()` via no-allocation fields like this:

https://github.com/juj/wasm_webgpu/blob/ddf33db3b8c625061b97f89a43a7bf23b2fa9826/lib/lib_webgpu.js#L860-L875

Reply:
Thanks @juj for jumping into this thread. I don't see `wgpuAdapterInfoFreeMembers` in juj/wasm_webgpu. Is that intended?

Reply:
Yeah, since the memory is caller allocated, there is no need to call a separate function to free up memory.

Reply:
I can raise the idea of using arrays instead of having a Free step with other webgpu.h folks. For now let's stick with what we have written down.